service/raft/raft_address_map.cc

```diff
@@ -0,0 +1,191 @@
+    if (app_state_ptr) {
+        raft::server_id id(utils::UUID(app_state_ptr->value));
+        add_or_update_entry(id, endpoint);
+    }
+    return make_ready_future<>();
+}
+
+void raft_address_map::unsubscribe_from_fd_events(direct_failure_detector::failure_detector&) {
+    _direct_fd_subscription.reset();
+}
+
+
+future<> raft_address_map::on_join(gms::inet_address endpoint, gms::endpoint_state ep_state) {
+
+    rslog.debug("raft_address_map::on_join {} {}", endpoint, ep_state);
+    return on_endpoint_change(endpoint, ep_state);
+}
+
+
+future<> raft_address_map::before_change(gms::inet_address endpoint,
+        gms::endpoint_state current_state, gms::application_state new_statekey,
+        const gms::versioned_value& newvalue)  {
+
+    // Raft server ID never changes - do nothing
+    rslog.debug("raft_address_map::before_change {} {} {}", endpoint, current_state, new_statekey);
+    return make_ready_future<>();
+}
+
+future<> raft_address_map::on_change(gms::inet_address endpoint,
+    gms::application_state state, const gms::versioned_value& value) {
+
+    // Raft server ID never changes - do nothing
+    rslog.debug("raft_address_map::on_change {} {}", endpoint, state);
+    return make_ready_future<>();
+}
+
+future<> raft_address_map::on_alive(gms::inet_address endpoint, gms::endpoint_state ep_state) {
+
+    rslog.debug("raft_address_map::on_alive{} {}", endpoint, ep_state);
+    return on_endpoint_change(endpoint, ep_state);
+}
+
+future<> raft_address_map::on_dead(gms::inet_address endpoint, gms::endpoint_state state) {
+    rslog.debug("raft_address_map::on_dead{} {}", endpoint, state);
+    return make_ready_future<>();
+}
+
+future<> raft_address_map::on_remove(gms::inet_address endpoint) {
+    // The mapping is removed when the server is removed from
+    // Raft configuration, not when it's dead or alive, or
+    // removed
+    rslog.debug("raft_address_map::on_remove {}", endpoint);
+    return make_ready_future<>();
+}
+
+future<>
+raft_address_map::on_restart(gms::inet_address endpoint, gms::endpoint_state ep_state) {
+
+    rslog.debug("raft_address_map::on_restart{} {}", endpoint, ep_state);
+    return on_endpoint_change(endpoint, ep_state);
+}
+
+future<>
+raft_address_map::subscribe_to_fd_events(direct_failure_detector::failure_detector& direct_fd) {
+
+    _direct_fd_subscription.emplace(co_await direct_fd.register_listener(*_direct_fd_proxy,
+        direct_fd_clock::base::duration{std::chrono::seconds{1}}.count()));
+
+    for (const auto& it : _set) {
+        if (it.expiring() || it._addr == gms::inet_address{}) {
+            continue;
+            notify_pinger(it._addr, true);
+        }
+    }
+}
+
+void
+raft_address_map::add_or_update_entry(raft::server_id id, gms::inet_address addr) {
+    bool is_expiring = this->raft_address_map_t<seastar::lowres_clock>::add_or_update_entry(id, addr);
+    if (!is_expiring) {
```

Comment:
I don't understand why only modify pinger if the entry is expiring. I think it may make sense to detect liveness of nodes that we didn't yet mark as permanent. Also, this is inconsistent with `set_expiring_flag` which notifies pinger.

Reply:
I think you got it reversed. We don't modify if the entry is expiring.

We only need to track liveness of raft configuration members. All such members are marked non-expiring. As soon as we learn which address to ping for such a node, we notify the pinger to track it.